src/goto-instrument/contracts/contracts.cpp

```diff
@@ -920,23 +1121,69 @@ void code_contractst::apply_loop_contract(
       if(inner == outer)
         continue;
 
-      if(loop_nesting_graph[outer].loop.contains(
-           loop_nesting_graph[inner].target))
-        loop_nesting_graph.add_edge(outer, inner);
+      if(loop_nesting_graph[outer].content.contains(
+           loop_nesting_graph[inner].head_target))
+      {
+        if(!loop_nesting_graph[outer].content.contains(
+             loop_nesting_graph[inner].end_target))
+        {
+          log.error()
+            << "Overlapping loops at:\n"
+            << loop_nesting_graph[outer].head_target->source_location()
+            << "\nand\n"
+            << loop_nesting_graph[inner].head_target->source_location()
+            << "\nLoops must be nested or sequential for contracts to be "
+               "enforced."
+            << messaget::eom;
+        }
```

Comment:
Why not covered by regression tests?

Reply:
I don't think we can write loops in C which would trigger this check. It can only be trigger if someone messes with the goto instructions directly and introduces some statements in the middle of a loop.

We should have this check in place though because our instrumentation code assumes this property and would be unsound if two loop bodies somehow overlap.